### PR TITLE
Framework compilers package for SDK consumption

### PR DESCRIPTION
**Pull request by jaredpar**
This is the Roslyn side of the fix for [msbuild 7832](https://github.com/dotnet/msbuild/issues/7832) It provides a NuPkg that contains _only_ the .NET Framework version of the Roslyn compilers. This matches, almost identically, what is inserted into the MSBuild install of Visual Studio.

The .NET SDK can now use this package to use a .NET Framework toolset when building C# and VB projects. This will resolve the analyzer inconsistency issues that come up when mixing .NET SDK and MSBuild Framework toolsets at different versions in CI.

Related
- https://github.com/dotnet/roslyn/issues/61252
- https://github.com/dotnet/msbuild/issues/7832

The package verification

### CODE REVIEW

**Comment by jaredpar** [comment]
@dotnet/roslyn-compiler PTAL
@marcpopMSFT, @rainersigwald FYI

File: src/NuGet/Microsoft.Net.Compilers.Toolset/Framework/Microsoft.Net.Compilers.Framework.Package.csproj [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+    <IsPackable>true</IsPackable>
+    <NuspecPackageId>Microsoft.Net.Compilers.Framework</NuspecPackageId>
+    <IncludeBuildOutput>false</IncludeBuildOutput>
+    <DevelopmentDependency>true</DevelopmentDependency>
+    <PackageDescription>
+      .NET Framework Compilers Toolset Package
+
+      Referencing this package will cause the project to be built using the C# and Visual Basic framework compilers contained in the package, as opposed to the version installed with MSBuild.
+
+      This package is an implementation detail of a .NET SDK feature. Directly referencing this package is not supported.
```
Comment (333fred) [inline src/NuGet/Microsoft.Net.Compilers.Toolset/Framework/Microsoft.Net.Compilers.Framework.Package.csproj:15]: Are we planning on pushing this package to nuget.org?

Comment (jaredpar) [reply src/NuGet/Microsoft.Net.Compilers.Toolset/Framework/Microsoft.Net.Compilers.Framework.Package.csproj:15]: Yes. The hope is that we can do this as unlisted.

Comment (333fred) [reply src/NuGet/Microsoft.Net.Compilers.Toolset/Framework/Microsoft.Net.Compilers.Framework.Package.csproj:15]: We might need to update the nuget publish tool to do that automatically then.

File: Roslyn.sln [hunk 2]
```diff
@@ -525,6 +525,8 @@ Project("{778DAE3C-4631-46EA-AA77-85C1314464D9}") = "Microsoft.CodeAnalysis.Visu
 EndProject
 Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "TestDiscoveryWorker", "src\Tools\TestDiscoveryWorker\TestDiscoveryWorker.csproj", "{8BC50AFF-1EBF-4E9A-AEBB-04F387AA800F}"
 EndProject
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Microsoft.Net.Compilers.Framework.Package", "src\NuGet\Microsoft.Net.Compilers.Toolset\Framework\Microsoft.Net.Compilers.Framework.Package.csproj", "{521ADC3E-CC15-414B-9356-D87C5BCF3A24}"
```
Comment (333fred) [inline Roslyn.sln:528]: Add to Compilers.snlf?

File: src/NuGet/Microsoft.Net.Compilers.Toolset/Framework/Microsoft.Net.Compilers.Framework.Package.csproj [hunk 3]
```diff
@@ -0,0 +1,47 @@
+﻿<!-- Licensed to the .NET Foundation under one or more agreements. The .NET Foundation licenses this file to you under the MIT license. See the LICENSE file in the project root for more information. -->
+<Project Sdk="Microsoft.NET.Sdk">
+  <PropertyGroup>
+    <TargetFramework>net472</TargetFramework>
+
+    <IsPackable>true</IsPackable>
+    <NuspecPackageId>Microsoft.Net.Compilers.Framework</NuspecPackageId>
```
Comment (jcouv) [inline src/NuGet/Microsoft.Net.Compilers.Toolset/Framework/Microsoft.Net.Compilers.Framework.Package.csproj:7]: >Microsoft.Net.Compilers.Framework [](http://example.com/codeflow?start=21&length=33)

Feels like "Toolset" should appear in the package name, no?

File: src/NuGet/Microsoft.Net.Compilers.Toolset/Framework/Microsoft.Net.Compilers.Framework.Package.csproj [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+      This package is an implementation detail of a .NET SDK feature. Directly referencing this package is not supported. 
+
+      $(RoslynPackageDescriptionDetails)
+    </PackageDescription>
+    <TargetsForTfmSpecificContentInPackage>$(TargetsForTfmSpecificContentInPackage);_GetFilesToPackage</TargetsForTfmSpecificContentInPackage>
+    <!-- Remove NU5128 once https://github.com/NuGet/Home/issues/8583 is fixed -->
+    <NoWarn>$(NoWarn);NU5100;NU5128</NoWarn>
+
+    <_DependsOn>InitializeDesktopCompilerArtifacts</_DependsOn>
+    <_DependsOn Condition="'$(TargetFramework)' == 'net6.0'">InitializeCoreClrCompilerArtifacts</_DependsOn>
```
Comment (jcouv) [inline src/NuGet/Microsoft.Net.Compilers.Toolset/Framework/Microsoft.Net.Compilers.Framework.Package.csproj:24]: >net6.0 [](http://example.com/codeflow?start=52&length=6)

I didn't follow the mix of `'$(TargetFramework)' == 'net6.0'` and `'$(TargetFramework)' == 'net472'` in this file. Can you help me understand?

**Review by jcouv — COMMENTED** [review]
Done with review pass (iteration 1)

**Comment by jaredpar** [comment]
Need to change this so it only loads when executing on MSBuild framework. Essentially `'$(MSBuildRuntimeType)' == 'Full'`

File: src/NuGet/Microsoft.Net.Compilers.Toolset/Framework/Microsoft.Net.Compilers.Framework.Package.csproj (hunk 1 above)
Comment (jaredpar) [reply src/NuGet/Microsoft.Net.Compilers.Toolset/Framework/Microsoft.Net.Compilers.Framework.Package.csproj:15]: Where is this tool so I can update it?

File: src/NuGet/Microsoft.Net.Compilers.Toolset/Framework/Microsoft.Net.Compilers.Framework.Package.csproj (hunk 4 above)
Comment (jaredpar) [reply src/NuGet/Microsoft.Net.Compilers.Toolset/Framework/Microsoft.Net.Compilers.Framework.Package.csproj:24]: Deleted the `net6.0` portion.

File: Roslyn.sln (hunk 2 above)
Comment (jaredpar) [reply Roslyn.sln:528]: done

File: azure-pipelines.yml [hunk 5]
```diff
@@ -296,27 +296,27 @@ stages:
...
-      - script: eng/test-build-correctness.cmd -configuration Release -enableDumps
-        displayName: Build - Validate correctness
+      - template: eng/pipelines/build-bootstrap.yml
+        parameters:
+          bootstrapToolset: 
 
-      - template: eng/pipelines/publish-logs.yml
+  - job: Correctness_Bootstrap_Build_Framework
+    dependsOn: Determine_Changes
+    condition: or(ne(variables['Build.Reason'], 'PullRequest'), eq(dependencies.Determine_Changes.outputs['SetPathVars_compilers.containsChange'], 'true'))
```
Comment (jaredpar) [inline azure-pipelines.yml:305]: Going to change this condition to only run this bootstrap on CI. Double bootstrap on PR is pretty pointless. But this is an important test cause it's a customer scenario.

File: src/NuGet/Microsoft.Net.Compilers.Toolset/Framework/Microsoft.Net.Compilers.Framework.Package.csproj (hunk 1 above)
Comment (333fred) [reply src/NuGet/Microsoft.Net.Compilers.Toolset/Framework/Microsoft.Net.Compilers.Framework.Package.csproj]: https://github.com/dotnet/roslyn-tools/blob/main/src/dotnet-roslyn-tools/NuGet/NuGetPublish.cs

File: src/NuGet/Microsoft.Net.Compilers.Toolset/Framework/build/Microsoft.Net.Compilers.Toolset.Framework.props [hunk 6]
```diff
@@ -0,0 +1,12 @@
+﻿<!-- Licensed to the .NET Foundation under one or more agreements. The .NET Foundation licenses this file to you under the MIT license. See the LICENSE file in the project root for more information. -->
+<Project>
+
+  <!-- 
+    This packag is meant to override the compiler when running on .NET Framework MSBuild. When running
```
Comment (jjonescz) [inline src/NuGet/Microsoft.Net.Compilers.Toolset/Framework/build/Microsoft.Net.Compilers.Toolset.Framework.props:5]:
```suggestion
    This package is meant to override the compiler when running on .NET Framework MSBuild. When running
```

File: src/NuGet/Microsoft.Net.Compilers.Toolset/Framework/build/Microsoft.Net.Compilers.Toolset.Framework.props [hunk 7]
```diff
@@ -0,0 +1,12 @@
...
+    inside of dotnet build it should be a no-op.
+  -->
+  <Import
+    Project="$(MSBuildThisFileDirectory)Microsoft.Net.Compilers.Toolset.Framework.Coreprops"
```
Comment (jjonescz) [inline src/NuGet/Microsoft.Net.Compilers.Toolset/Framework/build/Microsoft.Net.Compilers.Toolset.Framework.props:9]:
```suggestion
    Project="$(MSBuildThisFileDirectory)Microsoft.Net.Compilers.Toolset.Framework.Core.props"
```

File: src/NuGet/Microsoft.Net.Compilers.Toolset/Framework/build/Microsoft.Net.Compilers.Toolset.Framework.props (hunk 6 above)
Comment (jaredpar) [reply src/NuGet/Microsoft.Net.Compilers.Toolset/Framework/build/Microsoft.Net.Compilers.Toolset.Framework.props:5]: hahah ... yeah i'm terrible at the spelling thing.

File: src/NuGet/Microsoft.Net.Compilers.Toolset/Framework/build/Microsoft.Net.Compilers.Toolset.Framework.props (hunk 7 above)
Comment (jaredpar) [reply src/NuGet/Microsoft.Net.Compilers.Toolset/Framework/build/Microsoft.Net.Compilers.Toolset.Framework.props:9]: damn i'm bad.

**Comment by jaredpar** [comment]
FYI: added a README.md to help explain the rationale for why there are three different packages here that largely do the same thing.

File: src/NuGet/Microsoft.Net.Compilers.Toolset/Framework/Microsoft.Net.Compilers.Framework.Package.csproj (hunk 1 above)
Comment (jaredpar) [reply src/NuGet/Microsoft.Net.Compilers.Toolset/Framework/Microsoft.Net.Compilers.Framework.Package.csproj]: Okay will update that once we merge.

**Review by 333fred — APPROVED** [review]

File: src/NuGet/Microsoft.Net.Compilers.Toolset/Framework/Microsoft.Net.Compilers.Toolset.Framework.Package.csproj [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+      This package is an implementation detail of a .NET SDK feature. Directly referencing this package is not supported. 
+
+      $(RoslynPackageDescriptionDetails)
+    </PackageDescription>
+    <TargetsForTfmSpecificContentInPackage>$(TargetsForTfmSpecificContentInPackage);_GetFilesToPackage</TargetsForTfmSpecificContentInPackage>
+    <!-- Remove NU5128 once https://github.com/NuGet/Home/issues/8583 is fixed -->
+    <NoWarn>$(NoWarn);NU5100;NU5128</NoWarn>
+
+    <_DependsOn>InitializeDesktopCompilerArtifacts</_DependsOn>
+  </PropertyGroup>
```
Comment (jjonescz) [inline src/NuGet/Microsoft.Net.Compilers.Toolset/Framework/Microsoft.Net.Compilers.Toolset.Framework.Package.csproj:24]: Should we exclude this package from source build?

Comment (jaredpar) [reply src/NuGet/Microsoft.Net.Compilers.Toolset/Framework/Microsoft.Net.Compilers.Toolset.Framework.Package.csproj:24]: Actually yes we should.

**Review by jjonescz — APPROVED** [review]

**Pull request merged**